Zero value `MapValue` causes panic in `Equal` method

### Module version
<!---
Inspect your go.mod as below to find the version, and paste the result between the ``` marks below.

go list -m github.com/hashicorp/terraform-plugin-framework/...

If you are not running the latest version of the framework, please try upgrading
because your bug may have already been fixed.
-->

```
v1.6.1
```

### Expected Behavior
<!--
What should have happened?
-->
`Equal` handles a nil `elementType`. I'm not sure if the receiver having a nil type should always return `false`, or if an input that also has a nil type and equivalent state should technically match.

### Actual Behavior
<!--
What actually happened?
-->
Panic.

### Steps to Reproduce

1. Add a test case like the following to `TestMapValueEqual`.
```go
    "zero-null": {
        receiver: MapValue{},
        input:    NewMapNull(StringType{}),
        expected: false,
    },
```
2. Observe panic.
```console
% go test -v ./types/basetypes/... -run=TestMapValueEqual/zero-null
=== RUN   TestMapValueEqual
=== PAUSE TestMapValueEqual
=== CONT  TestMapValueEqual
=== RUN   TestMapValueEqual/zero-null
=== PAUSE TestMapValueEqual/zero-null
=== CONT  TestMapValueEqual/zero-null
--- FAIL: TestMapValueEqual (0.00s)
    --- FAIL: TestMapValueEqual/zero-null (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x20 pc=0x104e0eba0]

goroutine 19 [running]:
testing.tRunner.func1.2({0x104f481a0, 0x1051653c0})
        /opt/homebrew/Cellar/go/1.21.7/libexec/src/testing/testing.go:1545 +0x1c4
testing.tRunner.func1()
        /opt/homebrew/Cellar/go/1.21.7/libexec/src/testing/testing.go:1548 +0x360
panic({0x104f481a0?, 0x1051653c0?})
        /opt/homebrew/Cellar/go/1.21.7/libexec/src/runtime/panic.go:914 +0x218
github.com/hashicorp/terraform-plugin-framework/types/basetypes.MapValue.Equal({0x0?, {0x0?, 0x0?}, 0x7d?}, {0x104f9cb20?, 0x14000142240?})
        /Users/jaredbaker/development/terraform-plugin-framework/types/basetypes/map_value.go:259 +0x50
github.com/hashicorp/terraform-plugin-framework/types/basetypes.TestMapValueEqual.func1(0x14000124b60)
        /Users/jaredbaker/development/terraform-plugin-framework/types/basetypes/map_value_test.go:601 +0x58
testing.tRunner(0x14000124b60, 0x14000126bc0)
        /opt/homebrew/Cellar/go/1.21.7/libexec/src/testing/testing.go:1595 +0xe8
created by testing.(*T).Run in goroutine 18
        /opt/homebrew/Cellar/go/1.21.7/libexec/src/testing/testing.go:1648 +0x33c
FAIL    github.com/hashicorp/terraform-plugin-framework/types/basetypes 0.447s
FAIL
```

### References
Relates https://github.com/hashicorp/terraform-provider-aws/pull/36437

--- a/.changes/unreleased/BUG FIXES-20240319-085739.yaml
+++ b/.changes/unreleased/BUG FIXES-20240319-085739.yaml
@@ -1,0 +1,6 @@
+kind: BUG FIXES
+body: 'types/basetypes: Prevented panic in the `MapValue` types `Equal` method when
+  the receiver has a nil `elementType`'
+time: 2024-03-19T08:57:39.837509-04:00
+custom:
+  Issue: "961"

--- a/.changes/unreleased/BUG FIXES-20240319-085908.yaml
+++ b/.changes/unreleased/BUG FIXES-20240319-085908.yaml
@@ -1,0 +1,6 @@
+kind: BUG FIXES
+body: 'types/basetypes: Prevented panic in the `ListValue` types `Equal` method when
+  the receiver has a nil `elementType`'
+time: 2024-03-19T08:59:08.576502-04:00
+custom:
+  Issue: "961"

--- a/.changes/unreleased/BUG FIXES-20240319-085928.yaml
+++ b/.changes/unreleased/BUG FIXES-20240319-085928.yaml
@@ -1,0 +1,6 @@
+kind: BUG FIXES
+body: 'types/basetypes: Prevented panic in the `SetValue` types `Equal` method when
+  the receiver has a nil `elementType`'
+time: 2024-03-19T08:59:28.116063-04:00
+custom:
+  Issue: "961"

--- a/types/basetypes/list_value.go
+++ b/types/basetypes/list_value.go
@@ -258,6 +258,11 @@ func (l ListValue) Equal(o attr.Value) bool {
 		return false
 	}
 
+	// A list with no elementType is an invalid state
+	if l.elementType == nil || other.elementType == nil {
+		return false
+	}
+
 	if !l.elementType.Equal(other.elementType) {
 		return false
 	}

--- a/types/basetypes/list_value_test.go
+++ b/types/basetypes/list_value_test.go
@@ -618,6 +618,21 @@ func TestListValueEqual(t *testing.T) {
 			input:    nil,
 			expected: false,
 		},
+		"zero-null": {
+			receiver: ListValue{},
+			input:    NewListNull(StringType{}),
+			expected: false,
+		},
+		"zero-zero": {
+			receiver: ListValue{},
+			input:    ListValue{},
+			expected: false,
+		},
+		"null-zero": {
+			receiver: NewListNull(StringType{}),
+			input:    ListValue{},
+			expected: false,
+		},
 	}
 	for name, test := range tests {
 		name, test := name, test

--- a/types/basetypes/map_value.go
+++ b/types/basetypes/map_value.go
@@ -265,6 +265,11 @@ func (m MapValue) Equal(o attr.Value) bool {
 		return false
 	}
 
+	// A map with no elementType is an invalid state
+	if m.elementType == nil || other.elementType == nil {
+		return false
+	}
+
 	if !m.elementType.Equal(other.elementType) {
 		return false
 	}

--- a/types/basetypes/map_value_test.go
+++ b/types/basetypes/map_value_test.go
@@ -621,6 +621,21 @@ func TestMapValueEqual(t *testing.T) {
 			input:    nil,
 			expected: false,
 		},
+		"zero-null": {
+			receiver: MapValue{},
+			input:    NewMapNull(StringType{}),
+			expected: false,
+		},
+		"zero-zero": {
+			receiver: MapValue{},
+			input:    MapValue{},
+			expected: false,
+		},
+		"null-zero": {
+			receiver: NewMapNull(StringType{}),
+			input:    MapValue{},
+			expected: false,
+		},
 	}
 	for name, test := range tests {
 		name, test := name, test

--- a/types/basetypes/set_value.go
+++ b/types/basetypes/set_value.go
@@ -258,6 +258,11 @@ func (s SetValue) Equal(o attr.Value) bool {
 		return false
 	}
 
+	// A set with no elementType is an invalid state
+	if s.elementType == nil || other.elementType == nil {
+		return false
+	}
+
 	if !s.elementType.Equal(other.elementType) {
 		return false
 	}

--- a/types/basetypes/set_value_test.go
+++ b/types/basetypes/set_value_test.go
@@ -871,6 +871,21 @@ func TestSetValueEqual(t *testing.T) {
 			input:    nil,
 			expected: false,
 		},
+		"zero-null": {
+			receiver: SetValue{},
+			input:    NewSetNull(StringType{}),
+			expected: false,
+		},
+		"zero-zero": {
+			receiver: SetValue{},
+			input:    SetValue{},
+			expected: false,
+		},
+		"null-zero": {
+			receiver: NewSetNull(StringType{}),
+			input:    SetValue{},
+			expected: false,
+		},
 	}
 	for name, test := range tests {
 		name, test := name, test

